### DISCUSSION TITLE
linter: import sorting and grouping

### DISCUSSION THREAD

**Discussion by LIMPIX31** [oxc-project/oxc#2858] [Feature Request]
Need rule for lexicographical sorting and grouping of imports

## Details
* https://github.com/oxc-project/oxc/issues/2713#issuecomment-2009348403
* https://github.com/oxc-project/oxc/issues/2713#issuecomment-2013693288

**Comment by Boshen** [comment]
I'd say that reporting whether the imports are sorted or not is kind of useless, I wonder if this should be a codemod, or a linter rule with `--fix`.

**Comment by LIMPIX31** [comment]
> I'd say that reporting whether the imports are sorted or not is kind of useless, I wonder if this should be a codemod, or a linter rule with `--fix`.

Yes, there should be a `--fix` option for that rule.

**Comment by Boshen** [comment]
Is there an existing eslint plugin for this? I'll make a decision if I have more contexts.

**Comment by LIMPIX31** [comment]
> Is there an existing eslint plugin for this? I'll make a decision if I have more contexts.

https://github.com/lydell/eslint-plugin-simple-import-sort

**Comment by Dunqing** [comment]
The import plugin has `order` rule do the same thing https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md

**Comment by XantreDev** [comment]
> The import plugin has `order` rule do the same thing [import-js/eslint-plugin-import@`main`/docs/rules/order.md](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md?rgh-link-date=2024-03-18T02%3A40%3A03Z)

I think [eslint-plugin-simple-import-sort](https://github.com/lydell/eslint-plugin-simple-import-sort) just has good defaults. Oxc can have both configurable rule and with good default

**Comment by XantreDev** [comment]
import/order is doable but too complex. Has too much of configuration, require support. 
[import/order](https://github.com/import-js/eslint-plugin-import/blob/main/src/rules/order.js#L328) ~ 800 LOC
[simple-import-sort](https://github.com/lydell/eslint-plugin-simple-import-sort/blob/main/src/imports.js) ~ 200 LOC (it's not actually true, there are a lot of shared code)

I will continue research about porting of simple-import-sort

Also `import/order` [spams](https://github.com/lydell/eslint-plugin-simple-import-sort/#how-do-i-use-this-with-dprint:~:text=This%20plugin%20gives%20you%20a%20single%20error%20for%20each%20chunk%20of%20imports/exports%2C%20while%20import/order%20can%20give%20multiple%20(see%20Can%20I%20use%20this%20without%20autofix%3F%20for%20details).%20In%20other%20words%2C%20this%20plugin%20is%20noisier%20in%20terms%20of%20underlined%20lines%20in%20your%20editor%2C%20while%20import/order%20is%20noisier%20in%20terms%20of%20error%20count.) with errors, this is one of the reasons why it's harder to implement

`simple-import-sort` [uses](https://github.com/lydell/eslint-plugin-simple-import-sort/blob/f5dc3794643e72b8c4fe02af6d130dbd3ab5d203/src/shared.js#L613) tokens to sort ImportSpecifiers

**Comment by LIMPIX31** [comment]
> import/order is doable but too complex. Has too much of configuration, require support. [import/order](https://github.com/import-js/eslint-plugin-import/blob/main/src/rules/order.js#L328) ~ 800 LOC [simple-import-sort](https://github.com/lydell/eslint-plugin-simple-import-sort/blob/main/src/imports.js) ~ 200 LOC (it's not actually true, there are a lot of shared code)
> 
> I will continue research about porting of simple-import-sort
> 
> Also `import/order` [spams](https://github.com/lydell/eslint-plugin-simple-import-sort/#how-do-i-use-this-with-dprint:~:text=This%20plugin%20gives%20you%20a%20single%20error%20for%20each%20chunk%20of%20imports/exports%2C%20while%20import/order%20can%20give%20multiple%20(see%20Can%20I%20use%20this%20without%20autofix%3F%20for%20details).%20In%20other%20words%2C%20this%20plugin%20is%20noisier%20in%20terms%20of%20underlined%20lines%20in%20your%20editor%2C%20while%20import/order%20is%20noisier%20in%20terms%20of%20error%20count.) with errors, this is one of the reasons why it's hard to implement

I'm wondering if we can do import grouping relying on workspace (reading package.json and glob matching) or tsconfig.json paths rather than just the import source. 

This should configure well, because I'd like to see the `react` and `next` imports separate and at the very top, not mixed in with other modules.

**Comment by XantreDev** [comment]
> I'm wondering if we can do import grouping relying on workspace (reading package.json and glob matching) or tsconfig.json paths rather than just the import source.

What is the problem with import source?

**Comment by LIMPIX31** [comment]
> What is the problem with import source?

```ts
import { useEffect } from 'react'    // "react": "18.2.0"
import { cn } from '@repo/lib'       // "@repo/lib": "workspace:*"
import { Button } from 'components'  // "baseUrl": "src"
```
By looking at the string after `from` (import source) you can tell that it looks like an npm package. Whereas they can actually be categorized into: `package`, `monorepo package` and `path mapping` (from tsconfig paths).

The question is whether linter is allowed to read any files in the workspace.

**Comment by piotrpalek** [comment]
Did this die or is there a discussion somewhere else?

**Comment by oleasteo** [comment]
I find the [perfectionist/sort-imports](https://perfectionist.dev/rules/sort-imports) options to be less confusing than the import/order ones.
The [`customGroups`](https://perfectionist.dev/rules/sort-imports#customgroups) option is simpler and more straightforward than the [import/order](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md#pathgroups) `pathgroups`.

**Reply by XantreDev** [reply]
Yep it's much better then import/order

**Reply by shurane** [reply]
This was helpful! I just used the JS plugin system to import this:

```jsonc
{
  "jsPlugins": ["eslint-plugin-perfectionist"],
  "rules": {
    "perfectionist/sort-imports": "error"
  }
}
```

Now I'm able to use `oxlint --fix` and not really have to worry about figuring out the order by hand since it was struggling with a lot of sorts (members and groups). It is slower but at least it's not error prone and I don't need to look at anything manually.

**Comment by nicolastoulemont** [comment]
On our side, we were doing import ordering via eslint which was reaching 50min+ CICD runtimes. In the end we decide to move away from doing import ordering via our linter and instead rely on our formatter for that (prettier + [sort-import plugin](https://github.com/trivago/prettier-plugin-sort-imports)). It's regex based, "simple" and fairly fast.

**Comment by tmm1** [comment]
I'm interested in replacing these codemod/fix rules used by vscode eslint extension:

```
  "eslint.codeActionsOnSave.rules": [
    "unused-imports/no-unused-imports",
    "import-newlines/enforce"
  ],
```

https://www.npmjs.com/package/eslint-plugin-unused-imports
https://www.npmjs.com/package/eslint-plugin-import-newlines

**Comment by Boshen** [comment]
Import sorting will be implemented in the formatter https://github.com/oxc-project/oxc/discussions/13608
